### ISSUE TITLE
Windows Version

### ISSUE THREAD

**Issue by uppps** [kozec/sc-controller#215]
There is a WIP Windows branch but there have not been any progress in the last month. Can I expect a release anytime soon?

**Comment by kozec** [comment]
Nope, unless you know someone with at least some Windows coding experience :D

What I was trying to do was to get at least [SendInput](https://msdn.microsoft.com/en-us/library/ms646310(v=vs.85).aspx) level of "emulation", what would work badly and with no gamepad support. AFAIK, there is no way to actually emulate gamepad spare of replacing xinput.dll and/or some system libraries.

**Comment by Ryochan7** [comment]
There are two main options for emulating an Xbox 360 gamepad across the system on Windows. The old standby is using the Scp Virtual Bus Driver. That driver is not really being maintained anymore though. A newer driver that will eventually replace it is called ViGEm. One nice feature about that driver is that it will be capable of emulating more than just an Xbox 360 gamepad. It still has some stability problems so it is not ready for mass consumption yet. Hopefully those issues will be ironed out soon.

https://github.com/nefarius/ViGEm

**Comment by kozec** [comment]
Thanks, that sounds useful.

But just to be clear, Windows support is currently somewhere at very bottom of TODO list, somewhere bellow support for non-steam controllers, i18n and probably VR integration or something :D But contributions are always welcomed.

There is long list of stuff that needs to be solved in potential Windows port:
- :ballot_box_with_check: libusb is not exactly reliable under Windows and pyusb is not 100% compatible with Linux version. Later is partially solved at least for wired controller
- There is no uinput, so keyboard and mouse emulation has to be rewritten. This should be possible using winapi.
- There is no uinput, so gamepad emualtion has to be rewritten. Most likely using ViGEm that @Ryochan7  linked post above
- Windows uses different keycodes, button names etc, so conversion code has to be made to convert stuff like `Keys.KEY_LEFTMETA` to whatever numerical constant Windows uses for Winkey.
- GUI uses unix socket to communicate with daemon. And while it's easy to just use TCP here, it opens whole can of security issues along with possibility of everything crashing when something else uses same port
- :ballot_box_with_check: And when we are talking about daemon, there is no such thing on Windows. Or rather, proces can act as daemon without all magic needed on Linux. This is already partially fixed by converting daemon into normal "exe", but stuff like stopping and restarting currently does nothing.
- OSD (including on-screen keyboard) depends on X11 to create non-clickable, non-managed window. I'm not sure if same is even possible on Windows
- Similarly, stuff line restricting mouse to specific area depends on X11 as well. But adding support throug winapi shouldn't be hard
- And autoswitcher is X11-only as well.

I'll update that list when I remember rest :)

**Comment by BanCrash** [comment]
Well, thanks for at least having it on the todo list!!

**Comment by FunnyBunny14** [comment]
What's the current state of the windows branch? Basically, what I want to know is whether it can accurately detect the steam controller "raw" input because if it can, I can use it as a base for my own project.

**Comment by kozec** [comment]
Absolutely nothing is happening, I'm just writing down library that _may_ be usable as uinput replaceament.

http://vjoystick.sourceforge.net/site/index.php/dev216

**Comment by BanCrash** [comment]
Like the meme says, it's something!! Thanks for letting us know!

**Comment by uppps** [comment]
@kozec Any progress? There haven't been any new commits in the windows branch recently.

**Comment by faldor20** [comment]
I AM STILL WATCHING.... :) i have been for a long time, its awesome to see this might one day be a real thing, i would love to ditch the steam setup for my non steam games.

**Comment by kozec** [comment]
It kinda works already, as long as you don't need GUI :)

Main problem is this [ViGEM issue](https://github.com/ViGEm/ViGEmBus/issues/10) that can literally freeze Windows 7, but I can make another prerelease with xinput-compatible pad emulation at least for Win10.

**Comment by faldor20** [comment]
ahh i see, well considering most of the world is moving towards windows 10 anyway i think even if it never gets fixed dropping support for windows 7 while unfortunate is not totally unacceptable.

**Comment by dastious** [comment]
Win 7 user, seems to work but it only emulate ps4 pad, i cannot use the mouse pad or mix desktop option with mouse or keyboard. Also Desktop mode only work outside of game, inside game it stop working.
Also i need to launch many time the daemon, because it always return serial number can't be obtain.

Is there a way to get mouse pad ? Or something useful ? Also the camera pad for ps4 don't work :  relXY function is not here.
So for now the pad is not usable.

**Comment by kozec** [comment]
> Win 7 user, seems to work but it only emulate ps4 pad

That's intentional, and probably will stay like that for foreseeable future :( It's better alternative than crashing entire OS.

> Also i need to launch many time the daemon, because it always return serial number can't be obtain.

... but this is not. Those serials are working differently between "versions" of Steam Controller, but I thought I had it finally got it right. I'll add workaround for case when it can't be retrieved at all.

> Also the camera pad for ps4 don't work : relXY function is not here.

Probably good thing to prioritize, thanks for reminding me.

**Comment by pk044** [comment]
i can't move in-game(seems like the left stick works like the right one), do I have to remap the keys in some config? should it be done in one of the sccprofiles?

thanks for adding windows support!

**Comment by john-2000** [comment]
Firstly monster thanks! I have it working as a keyboard and mouse but I would like some help as I am not sure I am using it right. I have VIGEM installed and see a couple of profiles loaded when I start the daemon
how do I activate different profiles? (if I just wanted the SC as an generic / Xbox controller, could I just overwrite the currently loaded profile with the xbox one?)

I gather the OSD is not working yet (? or does it require some GTK dependencies on windows?

**Comment by Ryochan7** [comment]
Thank you for the work on the Windows port. I have tried it and the daemon can interact with ViGEm but there are issues with the controller state translation.

1. ABS_RY is inverted.
2. LS mapped to ABS_LX and ABS_LY causes some erratic behavior especially when the stick is moved almost due north. ABS_LY will get inverted only for a small region of the stick.
3. ABS_RZ seems to get stuck at about 3/4 pull sometimes.
4. The gamepad state seems to reset when there is no new activity on the Steam Controller. Buttons will not stay held down if there is no axis activity.


Since exclusive access to the Steam Controller cannot be obtained yet, I have been testing playing Turok 2 from GOG Galaxy.

**Comment by kozec** [comment]
> i can't move in-game(seems like the left stick works like the right one), do I have to remap the keys in some config? should it be done in one of the sccprofiles?

Yes, until GUI is there, editing sccprofile file in text editor should work.

> I start the daemon
how do I activate different profiles? (if I just wanted the SC as an generic / Xbox controller, could I just overwrite the currently loaded profile with the xbox one?)
> 
> I gather the OSD is not working yet (? or does it require some GTK dependencies on windows?

OSD menu should work, but I noticed it takes few seconds to load on Windows. Pressing steam button should spawn it.

@Ryochan7 thanks for report, I'll check on those points.

**Comment by Ryochan7** [comment]
I recorded a short video showing some of the issues that I was having. Hopefully that will be a useful reference.

https://www.youtube.com/watch?v=j_V0lwKFlTc

**Comment by kozec** [comment]
Okay, I believe I managed to fix all of that and put it into [prerelease number 3](https://github.com/kozec/sc-controller/releases/tag/v0.4.9.3).

**Comment by kozec** [comment]
> Basically, the warnings arise when you:
> 
> - Press the steam button when the overlay is already open.

That's by design, only one menu should be shown at time. Menus are "locking" buttons they are using and that mechanism is used to prevent multiple menus spawning at once.

> - Every time you select any of the overlay options.
> 
> The error is the same for all (Error 10054; dropping client),

That's just means that menu is finished and closed connection, so it's expected, but I should probably silence it.

**Comment by Ryochan7** [comment]
Just tried the latest pre-release and all of those issues have been fixed. Thank you.

**Comment by CriticalComposer** [comment]
My only issue currently is that sc-controller doesn't take over exclusive control of the device. So Lizard Mode or Steam Configs will activate _as well as_ the sc-controller profile. I figure that you're already aware of this but it is the largest hurdle for the program currently. Otherwise, excellent job on getting it to where its at. I'm looking forward to see the Windows version evolve into the full blown Steam Input replacement that the Linux version is.

**Comment by dan990** [comment]
Hello nice little program you have there. Im experiencing some problems mainly with windows defender and the steam controller not being recognized. Windows defender says it's harmfull, but I already made the exclusions. Previously on a pc with avast the overlay was working, but now with this new installation of windows when I press the guide buttom it doesnt show the little menu, and there are two errors on the scc daemon that say "failed to open device 2 on bus 1: no such device" and "failed to open '/win32/usb/1/2'".
I hope you can help me resolve my issues, maybe windows defender is preventing it still somehow to comunicate with the emulated devices? I installed vigem trough glosc and that does work but I need this program mainly for games that dont like the steam overlay. Thanks in advance.

**Comment by kozec** [comment]
[New pre-release added](https://github.com/kozec/sc-controller/releases/tag/v0.4.9.4)

It now includes GUI, which actually makes it even more experimental than one before :)

**Comment by kozec** [comment]
(also, I'm not ignoring @CriticalComposer message, it's just that AFAIK, this is just how Windows works. But I'll definitely take opportunity if I find that I'm mistaken)

**Comment by CriticalComposer** [comment]
Dang, that sucks. It kinda overrides some of the utility of sc-controller. You either end up with double inputs or you have to run a blank Desktop config in Steam, which technically works but the whole point of sc-controller (in my eyes) is to circumvent the Steam dependency. I'm not a huge Steam user (I prefer Itch and GOG) so I'm always looking for ways to ditch Steam while retaining the full capabilities of the Steam Controller.

**Comment by kozec** [comment]
> The recent update to ViGEm (https://github.com/ViGEm/ViGEmBus/releases/tag/v1.16.112) has changed from using Powershell to using a Windows setup wizard so it's a LOT easier to install for your average user.

Great, I changed page accordingly.

**Comment by dan990** [comment]
I have a problem, the program doesn't seem to hook with the controller i get the following errors: 

> [E] USB          Failed to open device 2 on bus 1: No such device (it may have been disconnected)
> [E] sc_by_cable  Failed to open '/win32/usb/1/2'

.
Before it kind of worked and the inputs where detected in windows devices, the emulated xbox pad recognized my inputs (this was in v0.4.9.3 on a previous windows 10 installation with avast) and the guide buttom worked and showed the modes but now when I run the scc-deamon.exe or sc_controller.exe nothing is happening, no response from the controller inputs (apart from lizard mode) on emulated xbox controller.

BTW, I have in this installation of windows a kind of phantom xbox controller that appears when I open the devices, it doesn't even show inputs to test. I have glosc and that works but do not know if  I somehow Flunked my installation of ViGEm.

**Comment by Bricky5** [comment]
On the Gyro per-axis control, i can't adjust any of the settings, they're all greyed out.

Same goes for the settings too.

**Comment by dan990** [comment]
The program only works when using the dongle, not when the controller is plugged directly.
I tried, and when putting new bateries and trying with the dongle the scc daemon suddenly comes to life.

I prefer to use a wired connection so I don't waste a lot of batteries.

**Comment by Bricky5** [comment]
Ok, more specificity
Using Dongle.
I'll try the batteries, but it's not the daemon. it's the GUI

**Comment by dan990** [comment]
Sorry about that, I meant the emulation, the GUI opens and all but there are no response from the controller.
Just tried it again connected and the daemon throws these two errors
```
[E] USB          Failed to open device 3 on bus 2: No such device (it may have been disconnected)
[E] sc_by_cable  Failed to open '/win32/usb/2/3'
```
The GUI controller symbol (the one that lights on green when emulating) is grey

Then connect via dongle and the daemon is responsive, all inputs I use are registered and the GUI controller symbol shows green.

I already tried using games when connected via dongle and it does work, all keybindings are responsive, high precision template works

In short, the program always worked, but only recognizes my controller via dongle, not when directly plugged in.

**Comment by Bricky5** [comment]
Ok, separate error from all the buttons in the GUI being greyed out. got it.

**Comment by kozec** [comment]
I'm sorry, I'm not entirely sure what are you reporting. But gyroscope is what's am implementing right now, so anything related to it doesn't work _yet_ and that's why it is grayed out in GUI.

**Comment by Bricky5** [comment]
Ah. Then what about the settings? why are they greyed out?

    On Tuesday, May 21, 2019, 2:05:22 PM EDT, kozec <notifications@github.com> wrote:  
 
 
I'm sorry, I'm not entirely sure what are you reporting. But gyroscope is what's am implementing right now, so anything related to it doesn't work yet and that's why it is grayed out in GUI.

**Comment by Bricky5** [comment]
Sorry about the mess of replying through email

To clarify:
The Global settings for everything are also greyed out

**Comment by dan990** [comment]
Its an experimental build for windows, I imagine that the linux version already has those features.

**Comment by Bricky5** [comment]
that's why this is in an issue titled "windows version"

**Comment by kozec** [comment]
> The Global settings for everything are also greyed out

Yes, that's same case. I should have probably write that somewhere.

**Comment by kazoooka** [comment]
Does Steam have to be running? Because when i exit Steam the controller goes into Lizard mode and is just kinda of a keyboard and mouse... I can't seem to get sc-controller to work... 

Everytime i open scc-daemon.exe i get these messages:

[I] Daemon       Starting SC Controller Daemon v0.4.5...
[I] Daemon       Created control socket C:\Users\xxx\AppData\Roaming\scc\daemon.socket
[I] Daemon       Initializing drivers...
[D] Mapper       Creating virtual devices...
[d] Mapper       Gamepad:  <ViGEm x360 Controller 0x0110baf0>
[d] Mapper       Mouse:    <Winapi mouse device 0x0127ea68>
[d] Mapper       keyboard: <Winapi keyboard device 0x0127eb98>
[I] Daemon       Activated default profile 'C:\Users\xxx\Downloads\sc-controller-0.4.9.4-win32-pre-release\share\default_profiles\Desktop.sccprofile'
[E] USB          Failed to open device 10 on bus 2: No such device (it may have been disconnected)
[E] sc_by_cable  Failed to open '/win32/usb/2/a'
[I] Daemon       Ready.

**Comment by kozec** [comment]
> Does Steam have to be running? Because when i exit Steam the controller goes into Lizard mode and is just kinda of a keyboard and mouse... I can't seem to get sc-controller to work...

No, running Steam will conflict with sc-controller.

**Comment by JPili** [comment]
I was having the same problem as rusagent and it took me a few hours to figure out but I realized that the problem might be that the sc-controller folder is not in the OS drive. My OS drive is C: but sc-controller was located in the D: drive. So I moved the folder to the C: drive and the problem was fixed!

You should add a warning that says the sc-controller folder should be in the OS drive. I'm using Windows 10, by the way.

**Comment by JPili** [comment]
Nevermind... It stopped working again for some reason... The problem isn't the folder location. It's something else.

**Comment by notNSANE** [comment]
Can confirm that this isn't working when wired, or that there's something wrong.

... and also, if we don't find a way to disable Lizard Mode all might be useless. Maybe if it won't conflict with Steam, then we could technically create a nulled config and load it up when we use SC Controller.

**Comment by kozec** [comment]
> ... and also, if we don't find a way to disable Lizard Mode all might be useless. Maybe if it won't conflict with Steam, then we could technically create a nulled config and load it up when we use SC Controller.

sc-controller disables lizard mode right after it gains access to steam controller, so once I manage to find reason why it can't get access to wired controller, that should be solved as well.

**Comment by JPili** [comment]
I have a bug that prevents me from opening the "Right Pad" window whenever there's anything set to it. It's openable if there's nothing set to it yet but as soon as I save any sort of binding to it. It's unable to be opened. Is that only happening to me?

**Comment by kozec** [comment]
[New release added](https://github.com/kozec/sc-controller/releases/tag/v0.4.9.5), should be fixing all bugs mentioned so far with exception of wired SC issues.

**Comment by CriticalComposer** [comment]
Just tested the new release and no matter what button I click on I get this error 

`Exception AttributeError: "Parameter instance has no attribute '_cparam'" in <object repr() failed> ignored
Traceback (most recent call last):
  File "C:\Users\SteamOS\Downloads\sc-controller-0.4.9.5-win32-pre-release\python/scc/gui/controller_widget.py", line 64, in on_click
    self.app.show_editor(self.id)
  File "C:\Users\SteamOS\Downloads\sc-controller-0.4.9.5-win32-pre-release\python/scc/gui/app.py", line 384, in show_editor
    ae = self.choose_editor(action, "", id)
  File "C:\Users\SteamOS\Downloads\sc-controller-0.4.9.5-win32-pre-release\python/scc/gui/binding_editor.py", line 162, in choose_editor
    e = ActionEditor(self.app, self.on_action_chosen)
  File "C:\Users\SteamOS\Downloads\sc-controller-0.4.9.5-win32-pre-release\python/scc/gui/action_editor.py", line 106, in __init__
    self.load_components()
  File "C:\Users\SteamOS\Downloads\sc-controller-0.4.9.5-win32-pre-release\python/scc/gui/action_editor.py", line 163, in load_components
    self.load_component(c)
  File "C:\Users\SteamOS\Downloads\sc-controller-0.4.9.5-win32-pre-release\python/scc/gui/action_editor.py", line 179, in load_component
    instance = cls(self.app, self)
  File "C:\Users\SteamOS\Downloads\sc-controller-0.4.9.5-win32-pre-release\python/scc/gui/ae/axis_action.py", line 45, in __init__
    self.circular_axis = MouseAction(Rels.REL_WHEEL)
  File "C:\Users\SteamOS\Downloads\sc-controller-0.4.9.5-win32-pre-release\python/scc/actions.py", line 169, in __init__
    pars = [ Parameter(v) for v in args ]
  File "C:\Users\SteamOS\Downloads\sc-controller-0.4.9.5-win32-pre-release\python/scc/actions.py", line 75, in __init__
    raise TypeError("Cannot convert %s" % (repr(value),))
TypeError: Cannot convert <Rels.REL_WHEEL: 1032>`

**Comment by kozec** [comment]
@CriticalComposer please, do right click on button that generates such error, select "copy" and paste results here. I can't find combination that triggers that issue.

**Comment by kozec** [comment]
@CriticalComposer are you using physical DS4, or emulating DS4? That first thing is not yet supported on Windows.

As for main bug, I have it replicated, so give me a moment.

**Comment by kozec** [comment]
[Another release](https://github.com/kozec/sc-controller/releases/tag/v0.4.9.6), fixing above breaking bug.

**Comment by theclarkmeister** [comment]
I can't add a second controller in the latest release.  The Daemon Returns the following log:

[I] Daemon       Starting SC Controller Daemon v0.4.5...
[I] Daemon       Created control socket C:\Users\Travis\AppData\Roaming\scc\daemon.socket
[I] Daemon       Initializing drivers...
[D] Mapper       Creating virtual devices...
[d] Mapper       Gamepad:  <ViGEm x360 Controller 0x011ded78>
[d] Mapper       Mouse:    <Winapi mouse device 0x011deea8>
[d] Mapper       keyboard: <Winapi keyboard device 0x010e0260>
[E] CemuHook     Bind failed: error 10048
[I] Daemon       Activated default profile 'C:\Users\Travis\Desktop\sc-controller-0.4.9.6-win32-pre-release\share\default_profiles\Desktop.sccprofile'
[I] Daemon       Ready.
[E] USB          sccd_usb_dev_hid_request: out: Input/Output Error
[E] sc           Failed to retrieve serial number
[E] USB          sccd_usb_dev_hid_request: out: Input/Output Error
[E] sc           Failed to retrieve serial number
[E] USB          sccd_usb_dev_hid_request: out: Input/Output Error
[E] sc           Failed to retrieve serial number
[E] USB          sccd_usb_dev_hid_request: out: Input/Output Error
[E] sc           Failed to retrieve serial number
[D] Daemon       Assigned default_mapper to <SC FC615205E8>
[I] Daemon       Controller added: <SC FC615205E8>
[E] USB          sccd_usb_dev_hid_request: out: Input/Output Error
[E] sc           Failed to retrieve serial number
[E] USB          sccd_usb_dev_hid_request: out: Input/Output Error
[E] sc           Failed to retrieve serial number
[E] USB          sccd_usb_dev_hid_request: out: Input/Output Error
[E] sc           Failed to retrieve serial number
[E] USB          sccd_usb_dev_hid_request: out: Input/Output Error
[E] sc           Failed to retrieve serial number
[E] USB          sccd_usb_dev_hid_request: out: Input/Output Error
[E] sc           Failed to retrieve serial number
[E] Daemon       Cannot add controller with ID '(null)', duplicate ID.
[D] sc_dongle    Failed to add controller to daemon

Turning both contoller off and back on will always result in the first controller turned on being successfully added and the second will fail.  Thanks for your work on creating this fantastic tool!

**Comment by kozec** [comment]
[Added new release](https://github.com/kozec/sc-controller/releases/tag/v0.4.9.7) which should fix all issues with wired controller and IO errors.

**Comment by dan990** [comment]
The right pad trackball in joystick mouse and joystick move mode seem to not work in the game controllers as xbox axis...

**Comment by theclarkmeister** [comment]
Just following up on my previous post about not being able to add a second controller.  Here is the log from the daemon in the updated version:

[I] Daemon       Starting SC Controller Daemon v0.4.5...
[I] Daemon       Created control socket C:\Users\Travis\AppData\Roaming\scc\daemon.socket
[I] Daemon       Initializing drivers...
[D] Mapper       Creating virtual devices...
[d] Mapper       Gamepad:  <ViGEm x360 Controller 0x02960ea8>
[d] Mapper       Mouse:    <Winapi mouse device 0x0295da88>
[d] Mapper       keyboard: <Winapi keyboard device 0x02960fd8>
[E] CemuHook     Bind failed: error 10048
[I] Daemon       Activated default profile 'C:\Users\Travis\Desktop\sc-controller-0.4.9.7-win32-pre-release\share\default_profiles\Desktop.sccprofile'
[I] Daemon       Ready.
[D] Daemon       Assigned default_mapper to <SC FC61921F95>
[I] Daemon       Controller added: <SC FC61921F95>
[E] input_libusb sccd_input_libusb_hid_request: out: Input/Output Error
[E] sc           Failed to retrieve serial number
[E] input_libusb sccd_input_libusb_hid_request: out: Input/Output Error
[E] sc           Failed to retrieve serial number
[E] input_libusb sccd_input_libusb_hid_request: out: Input/Output Error
[E] sc           Failed to retrieve serial number
[E] Daemon       Cannot add controller with ID '(null)', duplicate ID.
[D] sc_dongle    Failed to add controller to daemon

Looks slightly better than before, but as you can see still a few issues going on.  Let me know if I can do anything else to help out and as always, thanks for your help and hard work on this lovely piece of software!

**Comment by kozec** [comment]
@theclarkmeister 

> [E] Daemon Cannot add controller with ID '(null)', duplicate ID.

This sounds like bug and reason. I've created #497 as separate issue of it.

**Comment by notNSANE** [comment]
Super late but here's my current report for Windows 10.

First, Lizard mode or Steam isn't being disabled so it causes a lot of issues right now. Second, I don't know if it's because lizard mode or what, but buttons are incorrectly mapped and games don't respond correctly. Not sure how to test this. 

For the record, there aren't any game controllers recognized on the Game Controllers control panel. But if I press the Guide button, it opens the SC-Controller green panel on bottom left.

**Comment by notNSANE** [comment]
As a separate issue, the scc-daemon.exe isn't closing correctly when exiting the app, so it's stuck until we kill the process.

**Comment by kozec** [comment]
[Added new release.](https://github.com/kozec/sc-controller/releases/tag/v0.4.9.8)

scc-daemon now uses hidapi and requests exclusive access, which (at least on my machine) prevents Steam from grabbing controller, switching it to Lizard mode and generally wrecking havoc.

**Comment by indoormatic** [comment]
Did you revert the changes from #461 by accident? I always had that problem but it was finally fixed in v0.4.9.7, now I'm getting Failed to open device errors again with v0.4.9.8.

**Comment by notNSANE** [comment]
I can confirm that the new release doesn't start Lizard mode on Windows! 

But the OSD is broken and doesn't work.

**Comment by kozec** [comment]
@indoormatic No, that shoudn't be happening. And OSD should work as well :( I'll check if I missed anything on another machine, but any details are welcome.

**Comment by kozec** [comment]
@indoormatic @notNSANE I can't reproduce neither of your problems. Can you try closing everything scc-related, starting `scc-daemon.exe`  and copy-pasting any error messages from console window it should open?

**Comment by notNSANE** [comment]
Got a lot of data.

**With Steam opened**, then starting scc-daemon.exe

`[E] input_hidapi Failed to open device \\?\hid#vid_28de&pid_1142&mi_01#8&1c2f7806&0&0000#{4d1e55b2-f16f-11cf-88cb-001111000030}: hid_open_path failed
[E] sc_dongle    Failed to open '/hidapi//?/hid#vid_28de&pid_1142&mi_01#8&1c2f7806&0&0000#{4d1e55b2-f16f-11cf-88cb-001111000030}'
[E] input_hidapi Failed to open device \\?\hid#vid_28de&pid_1142&mi_04#8&13bb7739&0&0000#{4d1e55b2-f16f-11cf-88cb-001111000030}: hid_open_path failed
[E] sc_dongle    Failed to open '/hidapi//?/hid#vid_28de&pid_1142&mi_04#8&13bb7739&0&0000#{4d1e55b2-f16f-11cf-88cb-001111000030}'
[E] input_hidapi Failed to open device \\?\hid#vid_28de&pid_1142&mi_02#8&33f3044b&0&0000#{4d1e55b2-f16f-11cf-88cb-001111000030}: hid_open_path failed
[E] sc_dongle    Failed to open '/hidapi//?/hid#vid_28de&pid_1142&mi_02#8&33f3044b&0&0000#{4d1e55b2-f16f-11cf-88cb-001111000030}'
[E] input_hidapi Failed to open device \\?\hid#vid_28de&pid_1102&mi_02#8&15dae965&0&0000#{4d1e55b2-f16f-11cf-88cb-001111000030}: hid_open_path failed
[E] sc_by_cable  Failed to open '/hidapi//?/hid#vid_28de&pid_1102&mi_02#8&15dae965&0&0000#{4d1e55b2-f16f-11cf-88cb-001111000030}'
[E] input_hidapi Failed to open device \\?\hid#vid_28de&pid_1142&mi_03#8&101bc689&0&0000#{4d1e55b2-f16f-11cf-88cb-001111000030}: hid_open_path failed
[E] sc_dongle    Failed to open '/hidapi//?/hid#vid_28de&pid_1142&mi_03#8&101bc689&0&0000#{4d1e55b2-f16f-11cf-88cb-001111000030}'
[I] Daemon       Ready.`

**With Steam closed**, at first it was working fine (apparently), but after pressing the Guide button I've got this: 

`[[[EEE]]]   mmmeeennnuuu___dddaaatttaaa...ccc      FFFaaaiiillleeeddd   tttoo o o opopepenen n ' 'D'D:D:\:\G\GaGamameme'e':': : N NoNo o s susucuchch h f fifililele e o oror r d didirierrceetcctotororyry
y
[
[E[E]E] ] O OSOSDSD D                             FFFaaaiiillleeeddd   tttooo   dddeeecccooodddeee   mmmeeennnuuu


[E] menu_data.c  Failed to open 'D:\Game': No such file or directory
[E] OSD          Failed to decode menu
[E] menu_data.c  Failed to open 'D:\Game': No such file or directory
[E] OSD          Failed to decode menu
[E] menu_data.c  Failed to open 'D:\Game': No such file or directory
[E] OSD          Failed to decode menu`

**It seems that the software is failing because there's a space in the path**. I confirmed this by moving it to the root directory of the partition then changing "sc-controller" to "sc controller".

Hope this helps! Waiting for the next release, meanwhile I'll keep the app on the root folder

**Comment by kozec** [comment]
Ok, that makes sense. Only one program can use controller now, so if Steam is started before SC-Controller, Steam is who's controlling. It not really possible to "steal" it, at least not without risking crashing it.

**Comment by kozec** [comment]
OSD issue should be fixed in [latest release](https://github.com/kozec/sc-controller/releases)

**Comment by notNSANE** [comment]
is a constant 10% CPU usage normal?

**Comment by elboletaire** [comment]
@notNSANE IMHO you should specify which CPU do you have, as you're complaining about a CPU issue. My usage is completely normal using an i7-7700

![imatge](https://user-images.githubusercontent.com/153305/68947479-d7401e80-07b5-11ea-8e42-c535a31a5c8f.png)

**Comment by notNSANE** [comment]
you're right! AMD Ryzen 1600

**Comment by notNSANE** [comment]
so I'm launching SCC via a command line on Windows. To close it, I close the window, but sometimes it doesn't close and it's completely stuck. Appears like this on Device Manager.

![wohGvvo](https://user-images.githubusercontent.com/7162375/69449104-d4ef3e80-0d38-11ea-87dc-5c4c774cd706.png)

**Comment by kozec** [comment]
@notNSANE It sounds like closing window kills scc-daemon instantly, without giving it any time to cleanup. Does same thing happens if you close it using "control-c" or through option in GUI?

**Comment by notNSANE** [comment]
it doesn't happen via GUI or using Ctrl-C.

I'm guessing it happens in stress situations, like, with lower RAM or some other processes hanging out... but that only happens after several hours of playing. Besides, I always have opened several browsers at all times, which also hog the PC a bit over time.

I'll try to reproduce it again later, but isn't there a way to prevent the window from being closed instantly? (I have no idea how these actually work).

**Comment by magjir** [comment]
Hi, is anything needed in order to run scc with GUI? When I try to run sc-controller.exe, it does nothing. But when I run only deamon, it works properly (even in games).

**Comment by m-vlad** [comment]
OSD still doesn't show up for me even with the latest version.

**Comment by magjir** [comment]
@m-vlad OSD shows properly for me (after pressing "steam" button on controller and waiting for few seconds. Bud GUI doesn't. Have you installed any python packages or are there any other prerequisites other than vigem and xbox driver as written in wiki? Thanks for answer. :)

**Comment by kozec** [comment]
@m-vlad Bluetooth is not supported yet, sorry.

@magjir Can you try starting that exe from command line? It may display some error message.

**Comment by magjir** [comment]
> 
> 
> @m-vlad Bluetooth is not supported yet, sorry.
> 
> @magjir Can you try starting that exe from command line? It may display some error message.

Here is screenshot. Unfortunately, running sc-controller from powershell still does nothing. :/
![image](https://user-images.githubusercontent.com/33799459/70387716-ba5bcd00-19a8-11ea-8c1b-df23bb98fd31.png)
 I am curious if it might have problem with central european symbols in my user folder path... But scc-deamon works fine and osd is drawn. 
![image](https://user-images.githubusercontent.com/33799459/70387732-01e25900-19a9-11ea-8e1b-d57638e07c47.png)

**Comment by kozec** [comment]
[Added new release.](https://github.com/kozec/sc-controller/releases/tag/v0.4.9.10)

@magjir I've added `scc-controller-console.exe` binary that, unlike one I told you to run before, actually stays attached to terminal and should display some error message. Please, try that one as well.

**Comment by magjir** [comment]
@kozec 
Here is screen of error message. :) 

![image](https://user-images.githubusercontent.com/33799459/71450558-71579700-2764-11ea-95c7-00c5526a37ab.png)

_And do you have a donation link? You are doing great job and I'd like to appreciate it with small thank you._

**Comment by kozec** [comment]
> And do you have a donation link? You are doing great job and I'd like to appreciate it with small thank you.

All my e-begging is handled by [this section of readme](https://github.com/kozec/sc-controller#like-what-im-doing) :)

> Could it be possible that in case of this charset error, configs would be stored in installation path?

Yes, but I'd rather have it to handle those cases correctly. If Czech can break it this much, I don't wanna think what it will do with cyrilic or kanji...

**Comment by ChrisNonyminus** [comment]
I posted this elsewhere already (see the feature request for a gamecube button layout setting), but would it be possible to have a "freeform touch menu"? Like, similar to a radial menu but you can place the buttons anywhere on the menu.
Thanks.

**Comment by Catcher40** [comment]
`I'm having a weird problem using the program since the last 2 updates, i can no longer get the settings to appear no matter how many times in click it and also the osd now refuses to exit. 

I'm running this on windows 10 x64 pro version 1903 (Build 18362.535), using ryzen 7 3700x oc@4.3Ghz, corsair vengeance ddr4 2 x 8 gb @ 3200mhz, asus rog x370-f MB installed on 240gb ssd.

The last version that the settings and osd worked for me was v0.4.9.8.

**Comment by kozec** [comment]
@ChrisNonyminus Technically it wouldn't be too hard, but I can't think of any way to specify button locations (both format and UI). I don't feel like implementing my own image editor...

@Catcher40 try running `sc-controller-console.exe` and check if it displays any errors.

**Comment by Catcher40** [comment]
this is what i get running sc-controller-console.exe

[I] GUI          Starting SC Controller GUI (forced console) v0.4.5...
[D] GUI          Initializing python...
[d] GUI          Python home: C:\sc-controller
[d] GUI          Python path: C:\sc-controller\python;C:\sc-controller\python\scc\lib;C:\sc-controller\lib\python2.7;C:\sc-controller\lib\python2.7\lib-dynload;C:\sc-controller\lib\python2.7\site-packages

(sc-controller:33228): GLib-GIO-WARNING **: 03:51:13.540: win32 session dbus binary not found: C:\sc-controller\\gdbus.exe
W UDataManager  enumerate_children_async returned no files
[d] SCCC         Connecting to 'C:\Users\steve\AppData\Roaming\scc\daemon.socket'...
[E] SCCC         Connection failed: error 10061
[d] SCCC         Connecting to 'C:\Users\steve\AppData\Roaming\scc\daemon.socket'...

**Comment by Catcher40** [comment]
Also if i then click on the settings option i then get

Traceback (most recent call last):
  File "C:\sc-controller\python/scc/gui/app.py", line 477, in on_mnuGlobalSettings_activate
    gs = GlobalSettings(self)
  File "C:\sc-controller\python/scc/gui/global_settings.py", line 81, in __init__
    self.load_settings()
  File "C:\sc-controller\python/scc/gui/global_settings.py", line 114, in load_settings
    self.load_colors()
  File "C:\sc-controller\python/scc/gui/global_settings.py", line 176, in load_colors
    theme = self.app.config.get("osd_color_theme", "None")
  File "C:\sc-controller\python/scc/config.py", line 104, in get
    raise KeyError("Invalid config key: %s" % (key,))
KeyError: u'Invalid config key: osd_color_theme'

**Comment by theclarkmeister** [comment]
Latest version is outputting the following in the sc-daemon:

[I] Daemon       Starting SC Controller Daemon v0.4.5...
[I] Daemon       Created control socket C:\Users\Travis\AppData\Roaming\scc\daemon.socket
[I] Daemon       Initializing drivers...
[D] Mapper       Creating virtual devices...
[d] Mapper       Gamepad:  <ViGEm x360 Controller 0x013094e0>
[d] Mapper       Mouse:    <Winapi mouse device 0x01309610>
[d] Mapper       keyboard: <Winapi keyboard device 0x0120ef98>
[E] CemuHook     Bind failed: error 10048
[I] Daemon       Activated default profile 'C:\Users\Travis\Desktop\sc-controller-0.4.9.10-win32-pre-release\share\default_profiles\Desktop.sccprofile'
[I] Daemon       Ready.
[d] SCCC         Connecting to 'C:\Users\Travis\AppData\Roaming\scc\daemon.socket'...
[d[]d D]a emSoCn C C        Accepted new connection
     Connected to daemon, version 0.4.5
[I] Daemon       Registered scc-osd-daemon
[E] input_hidapi sccd_input_hid_request: hid_send_feature_report failed: A device attached to the system is not functioning.
[E] sc           Failed to retrieve serial number
[E] input_hidapi sccd_input_hid_request: hid_send_feature_report failed: A device attached to the system is not functioning.
[E] sc           Failed to retrieve serial number
[D] Daemon       Assigned default_mapper to <SC FC615205E8>
[I] Daemon       Controller added: <SC FC615205E8>
[E] input_hidapi sccd_input_hid_request: hid_send_feature_report failed: A device attached to the system is not functioning.
[E] sc           Failed to retrieve serial number
[E] input_hidapi sccd_input_hid_request: hid_send_feature_report failed: A device attached to the system is not functioning.
[E] sc           Failed to retrieve serial number
[E] input_hidapi sccd_input_hid_request: hid_send_feature_report failed: A device attached to the system is not functioning.
[E] sc           Failed to retrieve serial number
[E] input_hidapi sccd_input_hid_request: hid_send_feature_report failed: A device attached to the system is not functioning.
[E] sc           Failed to retrieve serial number
[D] Mapper       Creating virtual devices...
[d] Mapper       Gamepad:  <ViGEm x360 Controller 0x012119f0>
[d] Mapper       Mouse:    <Winapi mouse device 0x01221b38>
[d] Mapper       keyboard: <Winapi keyboard device 0x01221c68>
[I] Daemon       Activated default profile 'C:\Users\Travis\Desktop\sc-controller-0.4.9.10-win32-pre-release\share\default_profiles\Desktop.sccprofile'
[D] Daemon       (Re)using mapper 01220b50 for <SC FC61921F95>
[I] Daemon       Controller added: <SC FC61921F95>

I also notice that I can only change the profile for the first controller I connect.  Anything past that will always use the desktop profile.  Pushing the guide button on any connected controller opens up the profile switcher, but will only respond to the first controller connected.  Thanks for your continued work!

**Comment by Catcher40** [comment]
Is the windows version no longer being worked on as its been almost 11 weeks since the last update, i know this isn't a bug report but i didn't know where else to ask this.

**Comment by kozec** [comment]
Still alive, just busier than I expected. I've managed to add support for non-steam controllers and currently porting / re-adjusting GUI for its settings. What's left is gesture recognition and then port is more-or-less done.

**Comment by 8BitCerberus** [comment]
Finally getting to try the Windows version proper, and hot damn @kozec ! It's a thing of beauty!

I have noticed a few minor issues, the main issue is the OSD, it comes up when I hit the Steam button, but there doesn't seem to be any way to dismiss it if I happened to accidentally summon it. Perhaps a toggle, if it's already running instead of it trying to run it again and throwing an error in the deamon console, it could instead just send the OSD away? 

Also when I browse to all configs and select a config, it gets properly added but the OSD sub-menu for selecting the config doesn't go away. It stays on screen, on top of everything, until I kill the daemon process. The main OSD does go away, for example, if I summon it and then choose to turn off the controller, or choose a template from the main OSD window. Just seems to be the sub-menu one that gets stuck.

The other minor issue is with the daemon not closing out when I close the GUI, although that may be intentional now that I think about it, as a way to close the GUI but still have the controller mapping in place handled by the daemon. But even closing from the menu (not just hitting the X) the daemon keeps running along with the console host. Is there another to close the daemon without resorting to killing the task?

**Comment by kozec** [comment]
Thanks, I'm glad it works at least partially :)

> I have noticed a few minor issues, the main issue is the OSD, it comes up when I hit the Steam button, but there doesn't seem to be any way to dismiss it if I happened to accidentally summon it.

"B" button should dismiss menu, but it sounds like there is some issue with that. In fact, from your description, it sounds like closing any menu doesn't work at all. Can you try running `scc-daemon.exe` manually, so it spawns console and displays errors?

> But even closing from the menu (not just hitting the X) the daemon keeps running along with the console host. Is there another to close the daemon without resorting to killing the task?

This, too, shouldn't be happening. When you have that `scc-daemon.exe` running, try starting `sc-controller.exe` as well and use that menu option to stop the daemon. It sounds like there is something that keeps processes alive even after they are supposed to exit.

**Comment by 8BitCerberus** [comment]
Hmm, no errors in coming up in console. B does dismiss the top-level menu but not the sub-menu.

Here's the output:

> [I] Daemon       Starting SC Controller Daemon v0.4.5...
[I] Daemon       Created control socket C:\Users\[user]\AppData\Roaming\scc\daemon.socket
[I] Daemon       Initializing drivers...
[D] Mapper       Creating virtual devices...
[d] Mapper       Gamepad:  <ViGEm x360 Controller 0x012bece0>
[d] Mapper       Mouse:    <Winapi mouse device 0x011aee78>
[d] Mapper       keyboard: <Winapi keyboard device 0x012bee10>
[I] CemuHook     Created CemuHookUDP Motion Provider
[I] Daemon       Activated default profile 'G:\Utilities\Steam Controller\sc-controller\share\default_profiles\Desktop.sccprofile'
[d] SCCC         Connecting to 'C:\Users\[user]\AppData\Roaming\scc\daemon.socket'...
[I] Daemon       Ready.
[[dd]]  DSaCeCmCo n              A ccepted new connectionC
onnected to daemon, version 0.4.5
[I] Daemon       Registered scc-osd-daemon
[d] Daemon       Accepted new connection
[D] Daemon       Assigned default_mapper to <SC FC53903366>
[I] Daemon       Controller added: <SC FC53903366>
[I] Daemon       Activating profile 'G:/Utilities/Steam Controller/sc-controller/share/default_profiles/Desktop.sccprofile'
[I] Daemon       Activated profile 'G:/Utilities/Steam Controller/sc-controller/share/default_profiles/Desktop.sccprofile' on default_mapper
[I] Daemon       Activating profile 'G:/Utilities/Steam Controller/sc-controller/share/default_profiles/Desktop.sccprofile'
[I] Daemon       Activated profile 'G:/Utilities/Steam Controller/sc-controller/share/default_profiles/Desktop.sccprofile' on default_mapper
[I] Daemon       Activating profile 'C:/Users/[user]/AppData/Roaming/scc/profiles/Desktop (Gyro).sccprofile'
[I] Daemon       Activated profile 'C:/Users/[user]/AppData/Roaming/scc/profiles/Desktop (Gyro).sccprofile' on default_mapper
[d] css          Loading css_provider G:\Utilities\Steam Controller\sc-controller\share/osd_styles/Reloaded.gtkstyle.css
[d] Daemon       Source 20/STICK locked on <SC FC53903366> by client 011b7790
[d] Daemon       Source 9/A locked on <SC FC53903366> by client 011b7790
[d] Daemon       Source 11/B locked on <SC FC53903366> by client 011b7790
[d] OSD          Opening submenu 'G:\Utilities\Steam Controller\sc-controller\share\default_menus\Profiles.menu'
[I] Daemon       Activating profile 'G:\Utilities\Steam Controller\sc-controller\share\default_profiles\Desktop.sccprofile'
[I] Daemon       Activated profile 'G:\Utilities\Steam Controller\sc-controller\share\default_profiles\Desktop.sccprofile' on default_mapper
[I] Daemon       Activating profile 'G:/Utilities/Steam Controller/sc-controller/share/default_profiles/Desktop.sccprofile'
[I] Daemon       Activated profile 'G:/Utilities/Steam Controller/sc-controller/share/default_profiles/Desktop.sccprofile' on default_mapper
[I] Daemon       Activating profile 'G:/Utilities/Steam Controller/sc-controller/share/default_profiles/Desktop.sccprofile'
[I] Daemon       Activated profile 'G:/Utilities/Steam Controller/sc-controller/share/default_profiles/Desktop.sccprofile' on default_mapper


And disabling the emulation from the menu, then exiting sc-controller does properly end all processes. But just exiting sc-controller while the emulation is active keeps both sc-controller.exe in memory and tied to scc-daemon.exe. Alternatively running scc-daemon first then running sc-controller, then exiting sc-controller ends it's process but scc-daemon continues running still (which would be expected as they were run separately.)

**Comment by notNSANE** [comment]
while I recall reading that this was intended, I fully support that closing SC Controller GUI closes the daemon too. It's basic UX, not doing it is decades of behavior ignored.

**Comment by djismgaming** [comment]
Just tested with bunch of games and only the X axis is moving when using Mouse Like Joystick on the Right Pad.

Moving is very little (sensitivity helps in that regard) but no Y axis movement is registered from controller.

**Comment by 8BitCerberus** [comment]
I noticed the lack of Y axis too, but I was only testing on one game so far and I wasn't sure yet if it was sc-controller or Cemu.

**Comment by Ryochan7** [comment]
I am experiencing the lack of Y axis movement with Mouse-like Joystick as well. One other big problem with the current beta build is that the Rotate modifier does not work. Pretty much makes the left touchpad useless for me. The c branch version is coming along nicely overall.

I might send a pull request with a couple of changes that I have found to improve things.

**Comment by blablabla3k** [comment]
Hi kozec, thanks for develo;pment windows vers. and all.
My question - can i attach steam controller directly to notebook or PC without original blutooth controller adapter, use BLE protocol in steam controller?
When i can use it itself - your cs controller application dont recognized steam controller and controller work as mouse standart.
Help.

**Comment by kozec** [comment]
@blablabla3k Sorry, bluetooth support is not implemented yet. It's one of those things Windows does very differently from everything else, so I'm saving it for when everything else works :)

As for Y axis on Mouse Like Joystick, I'll try to get it fixed in next release.

**Comment by LelouBil** [comment]
Hey, just came here quickly to say that my antivirus (Avast) detected that scc-daemon and scc-osd-daemon were infected by "IDP.Generic". I don't know what could have caused this so i'm just reporting it.

**Comment by LelouBil** [comment]
Actually, it was just scc-osd-daemon, I re-downloaded it just to be sure and it's still detected. Here's a Virustotal report https://www.virustotal.com/gui/file/9ce174bb6d677903c57adb3e3279d0581d382da176b64c34a46b07204c290130/detection

**Comment by patho5** [comment]
@kozec this tool is AMAZING!  I hope you continue to develop it.  Would love to see support for Bluetooth added in the future, as others have mentioned, but I understand there are more pressing issues.  :)

I am encountering the same issue @8BitCerberus mentioned above, with the Profiles sub-menu getting stuck on-screen.  The only way I have found to clear it is to end the scc-daemon.exe process from Task Manager.  I tried your suggestion of running scc-daemon.exe manually and observing errors, but no errors were thrown.  I can provide logs if you'd like.

Two other issues I have noticed:
- Haptic feedback does not seem to work on either touchpad.  Regardless of feedback settings in the GUI there is no feedback felt on the controller.
- Settings menu does not open, line "KeyError: u'Invalid config key: osd_color_theme'" appears in console.  Full log available upon request.

Lastly a usage question: Is it possible to launch the daemon silently, without having to launch the GUI first?

Thanks again for creating this, I hope to see it continue to grow!

**Comment by elboletaire** [comment]
> The only way I have found to clear it is to end the scc-daemon.exe process from Task Manager. 

If you're using the UI, you can also disable and enable emulation again, just by pressing the green dot in the top-left side of the app. That's the way I do remove the buggy sub-menu from screen these days.

**Comment by Ryochan7** [comment]
Is there any plan to utilize HidGuardian in the Windows build in the future? As I have experienced with my time developing DS4Windows, the old Exclusive Access flag option for hiding a device is becoming more useless by the month. More software keeps getting updated that gets more aggressive with grabbing references to HID devices which causes Exclusive Access to fail. Two big culprits are Microsoft Edge and the GeForce Experience suite (specifically the In-Game portion).

Hiding the Steam Controller from SC Controller and Steam is possible using HidGuardian. However, whitelisting the SCC Daemon is not feasible right now as it does not support device hotplugging.

**Comment by kozec** [comment]
> Hiding the Steam Controller from SC Controller and Steam is possible using HidGuardian. However, whitelisting the SCC Daemon is not feasible right now as it does not support device hotplugging.

Hotpluging is definitelly something that should work in "final" release, it was just not priority so far. As for HidGuardian, I'll have to look into it.

**Comment by kozec** [comment]
Added [new pre-release](https://github.com/kozec/sc-controller/releases/tag/v0.4.9.11), with all bugs known so far hopefully fixed.

**Comment by Repawn** [comment]
Was trying this out for the first time and I'm having the Y axis issue with mouse-like joystick mentioned above. Mouse-like joystick also seems jittery when dragging slowly.

**Comment by lagseeing** [comment]
The import/export window fails to open.

**Comment by jamiethomaswhite** [comment]
Pad press (click) is not working on Windows (10 20H2), I am unable to edit either pad "Pressed" option. Right-clicking the pad actions and clicking the "Edit Pressed Action" context-menu option does nothing, the context-menu closes and no dialog window opens up. Whereas the right joystick for example, right-clicking and choosing the "Edit Pressed Action" context-menu option works and opens up the action editing dialog.

Another issue on Windows is that setting a name, or renaming, any action in the "Action Name" area, renders that item unusable. You can no longer edit the item at all, clicking it does nothing and only clearing all bindings, starting over and **not** naming/renaming at all is the only option at the moment. It's a valid workaround but it just makes it harder to tell at a glance what you mapped things to.

Obviously, the "Cancel Button" and "Confirm selection by releasing the button" don't work still in Windows. They just reset/untick every time you close out the dialog.


One last thing, it would be nice is if I could change the theme to dark mode on Windows. In other GTK3 apps you can apparently edit the theme in the code or make some ini's to force it, but being on Windows and not seeing the same options/folder structure I'm unsure of how to do it.

**Comment by babasior** [comment]
Hi,

First of all, many thanks for this amazing work !

I made different kind of test with Windows version 4.9.12    and i have some issues 
- 1. OSD Menu has a bug (from after release 4.9.9), i can't exit from submenu with B button (it was working previously). 
- 2. I can't record second controller (first controller is Steam Controller - Working fine) 8bitdo SN30 Pro : i can select it but after button mapping windows if i click on next nothing happen  .... Perhaps developpement of this part is not finished ?
- 3. Profile import/export menu is not working.

Questions 
do you know if rumble feedback is supported on games with steam controller (and how to select it ?) ?


Here my OSD menu script modification for 4.9.12 Default.menu file

[{
    "separator": true,
    "name": "Select profiles"
}, {
    "generator" : "profiles"
}, {
    "separator": true,
    "id": "separator-after-profile-list",
    "name": "Options"
}, {
    "action": "turnoff()",
    "id": "turnoff_item",
    "icon": "system/turn-off",
    "name": "Turn Controller OFF",
    "osd": true
}]

Could you launch         scc-daemon.exe     on same time than          scc-controller.exe    please ?
I made a batch script for that and it's working very well.

**Comment by babasior** [comment]
here wrong mapping detection for SN30 pro, as you can see with raw data   "rpad_y" is wrong ("rpad_x" is ok) and "stick_x" and "stick_y" are not centered correctly.

{
    "axes": {
        "72": {
            "axis": "ltrig",
            "max": 32767,
            "min": -32768
        },
        "73": {
            "axis": "rtrig",
            "max": 32767,
            "min": -32768
        },
        "1048576": {
            "axis": "stick_x",
            "deadzone": 3586,
            "max": 65535,
            "min": 0
        },
        "1048577": {
            "axis": "stick_y",
            "deadzone": 17688,
            "max": 0,
            "min": 65535
        },
        "1048578": {
            "axis": "rpad_x",
            "deadzone": 65538,
            "max": 65535,
            "min": 0
        },
        "1048579": {
            "axis": "rpad_y",
            "deadzone": 4,
            "max": 32767,
            "min": -32768
        }
    },
    "buttons": {
        "64": "B",
        "65": "A",
        "66": "C",
        "67": "Y",
        "68": "X",
        "70": "LB",
        "71": "RB",
        "74": "BACK",
        "75": "START",
        "77": "STICKPRESS",
        "78": "RPADPRESS"
    },
    "dpads": {},
    "emulate_c": false,
    "gui": {
        "background": "x360",
        "buttons": [
            "A",
            "B",
            "X",
            "Y",
            "BACK",
            "C",
            "START",
            "LB",
            "RB",
            "LT",
            "RT",
            "LG",
            "RG"
        ]
    }
}

**Comment by rusagent** [comment]
I get this error in the console and also the controller is not working as it should. It's still in desktop mode and the controller doesn't connect at all.


![image](https://user-images.githubusercontent.com/1582815/104849921-b230c980-58ec-11eb-8481-453cd9540caf.png)

(sc-controller:16620): GLib-GIO-WARNING **: 17:43:18.327: win32 session dbus binary not found: F:\sc-controller-0.4.9.11-win32-pre-release\\gdbus.exe
W UDataManager  enumerate_children_async returned no files
[d] SCCC         Connecting to 'C:\Users\flori\AppData\Roaming\scc\daemon.socket'...

**Comment by DFOXpro** [comment]
Things I found not working or crashing the UI/daemon:

- Import/Export Modal is not working
- Circular input on trackpad do nothing
- Actions with None params (like empty buttons on dpad) crash the ui on profile select via  dropdown or with daemon on overlay default menu

Also thanks for porting this to windows, playing vr with the circular input willbe like a dream ;D

**Comment by gelatinbomb** [comment]
@rusagent I have the same problem.
Also haptic feedback is not working correctly. It appears it works but when I try to adjust the haptics it stop working and after that I can't fix the haptics again on this profile.
I found when I change the haptics the value gets corrupted. Lets say I choose 4000, when I click ok the value gets changed to 4000000.
Another problem is when you apply a button to the fully press trigger it gets applied to the semi-pressed trigger.
Sadly for these problems the sc-controller console didn't show me anything.
I'm happy to see sc-controller on Windows. I just got my first steam controller so If you need help testing stuff I'm here!

**Comment by gelatinbomb** [comment]
I found that using values from 1 to 9 you can get the feedback working since the program will always add three zeros extras on the configuration file.
I find this on the code. Could this be the part causing the problem? It would explain why every number I add have like three extra zeros.
![image](https://user-images.githubusercontent.com/31792239/122793975-fdbf1900-d278-11eb-839f-06c7089c7aee.png)
The dpad settings can't get any haptics values btw. No matter what I try the haptics get lost.
There is a lot of weird stuff happening actually. If I manage to get good at coding I could give you a hand. But for now I'm bare-bones at coding, drivers, and windows stuff.

**Comment by DFOXpro** [comment]
> Another problem is when you apply a button to the fully press trigger it gets applied to the semi-pressed trigger.
I believe this also happens in the Linux build, it's like the full-press is ignoring the physical button for the triggers

**Comment by 14142000** [comment]
> If someone is still watching this and feels _really_ adventurous, there is now work-in-progress SCC that [should work on Windows](https://github.com/kozec/sc-controller/releases/tag/v0.4.9.1).

I am excited because steam input doesnt work in battle.net games and I would love more granular control of my xbox one gamepad

**Comment by Horizon-omirza** [comment]
I am facing the same issue as https://github.com/kozec/sc-controller/issues/596

I've reinstall vigembus, tried different usb ports for the dongle. Nothing seems to work.

**Comment by axm2** [comment]
I'm having the same issue https://github.com/kozec/sc-controller/issues/596

**Comment by rusagent** [comment]
> Sorry for asking this here but I stopped using and checking on this over a year ago because of no updates, has it been abandoned it's been about 2 years since the last update. the non-windows version seems to be still worked on though.

What are you even asking? xD

**Comment by Ryochan7** [comment]
> I'm having the same issue as #596 Tried reinstalling vigembus, trying different usb ports.

Looks like SC Controller cannot open any virtual device from the SC Dongle based on the report. IIRC the Dongle acts as a hub that can present 4 Steam Controller devices to the PC. I cannot remember if SC Controller tries to open controllers in Exclusive mode or not but I would assume that some other program already has an open handle on those devices so that is why SC Controller fails to open them. Devices can be opened already if you have the Dongle plugged in at Windows startup thanks to the Windows Store app service IIRC.

You will likely want to check out how to use the HidHide driver with SC Controller.

https://vigem.org/projects/HidHide/

**Comment by Ryochan7** [comment]
> Sorry for asking this here but I stopped using and checking on this over a year ago because of no updates, has it been abandoned it's been about 2 years since the last update. the non-windows version seems to be still worked on though.

Working on a different mapper currently. Maybe it will get somewhere someday. It is working okay for me but there are many bugs to work out and there is pretty much no usable GUI atm. I also keep stumbling over my own ideas. Gotten a lot of use out of my Steam Controller over the last 3 months.

https://gitlab.com/ryochan7/steam-controller-test/-/tree/action_ideas

Remember Me - Steam Controller
https://www.youtube.com/watch?v=-QvJoQMGupg

**Comment by axm2** [comment]
> > I'm having the same issue as #596 Tried reinstalling vigembus, trying different usb ports.
> 
> Looks like SC Controller cannot open any virtual device from the SC Dongle based on the report. IIRC the Dongle acts as a hub that can present 4 Steam Controller devices to the PC. I cannot remember if SC Controller tries to open controllers in Exclusive mode or not but I would assume that some other program already has an open handle on those devices so that is why SC Controller fails to open them. Devices can be opened already if you have the Dongle plugged in at Windows startup thanks to the Windows Store app service IIRC.
> 
> You will likely want to check out how to use the HidHide driver with SC Controller.
> 
> https://vigem.org/projects/HidHide/

That did the trick, thanks for the help 👍

**Comment by Ryochan7** [comment]
Got the code at the point where it is somewhat usable. Putting a copy of the current progress out there for people to test. Still a lot of possible binding options are not available in the GUI. All supported connection types work on my end at least. You will likely want to configure HidHide to hide the Steam Controller from Steam when you don't want to map with Steam Input. Please let me know if you are interested in further development. Might create a GitHub mirror repo if people want it.

One other thing to mention. Only Windows 10 and 11 would be support mainly due to the use of ViGEmBus 1.17.333.0. I can only test on Windows 10 due to the CPU in my PC not fitting Win11 minimum requirements.

Required Dependencies:
.NET 5 Runtime x64 - https://dotnet.microsoft.com/en-us/download/dotnet/thank-you/runtime-5.0.17-windows-x64-installer
ViGEmBus 1.17.333.0 - https://github.com/ViGEm/ViGEmBus/releases/download/setup-v1.17.333/ViGEmBusSetup_x64.msi
FakerInput - https://github.com/Ryochan7/FakerInput/releases/download/v0.1.0/FakerInput_0.1.0_x64.msi

Optional Tools:
HidHide - https://vigem.org/projects/HidHide/

Source:
https://gitlab.com/ryochan7/steam-controller-test/-/tree/action_ideas

Archive:
https://drive.google.com/file/d/1yf2noIjkwnS3sbpbJqxHSdWK3i7b4bR8/view?usp=sharing

![example_mapper_screen_20220605](https://user-images.githubusercontent.com/581968/172064947-bc3358c7-7ec3-4a5d-86f5-3715bd4638e7.png)

![example_mapper_profile_overview_20220605](https://user-images.githubusercontent.com/581968/172065778-1a083361-e343-4364-9eb6-adc38ed5c6d2.png)

**Comment by axm2** [comment]
Gave it a quick try with forza tonight and I found a little typo with the default config. https://gitlab.com/ryochan7/steam-controller-test/-/blob/action_ideas/SteamControllerTest/template_profiles/SteamController/XInput.json#L306 

I think we should continue discussion elsewhere seeing as this is a separate project. You can make a github mirror or we can talk on gitlab if theres an 'issues' section on there.

**Comment by Ryochan7** [comment]
Set up a GitHub repo with the current code. Changed a few items since yesterday in the action_ideas branch.

https://github.com/Ryochan7/SteamControllerTest

Thank you for spotting the trigger binding problem in that example profile. I'll get that changed.

**Comment by pattontim** [comment]
Should I be altering the PATH, PYTHONPATH or other variables in order to get the GUI running after building? Could someone post me their env vars of a correctly building setup with ninja -C build sc-controller ?

This is my incorrect setup
```
Python path configuration:
  PYTHONHOME = (not set)
  PYTHONPATH = 'C:\Python27\Lib'
  program name = 'C:\msys64\mingw32\bin\python.exe'
  isolated = 0
  environment = 1
  user site = 1
  import site = 1
  sys._base_executable = 'C:/msys64/mingw32/bin/python.exe'
  sys.base_prefix = 'C:/msys64/mingw32'
  sys.base_exec_prefix = 'C:/msys64/mingw32'
  sys.platlibdir = 'lib'
  sys.executable = 'C:/msys64/mingw32/bin/python.exe'
  sys.prefix = 'C:/msys64/mingw32'
  sys.exec_prefix = 'C:/msys64/mingw32'
  sys.path = [
    'C:\\Python27\\Lib',
    'C:/msys64/mingw32/lib/python39.zip',
    'C:/msys64/mingw32/lib/python3.9',
    'C:/msys64/mingw32/lib/python3.9/lib-dynload',
    '',
  ]
```

**Comment by pattontim** [comment]
libmicroutils is no longer offered, you can install it from a mirror repository (mingw-w64-i686-libmicroutils-4.3.0-1) but it might cause the build to fail.

`$ meson build
The Meson build system
Version: 0.62.2
Source dir: C:/Users/tim/Documents/sc-controller
Build dir: C:/Users/tim/Documents/sc-controller/build
Build type: native build
Project name: scc
Project version: 3.14
C compiler for the host machine: cc (gcc 12.1.0 "cc (Rev2, Built by MSYS2 project) 12.1.0")
C linker for the host machine: cc ld.bfd 2.38
Host machine cpu family: x86
Host machine cpu: x86
Library regex found: YES
Library m found: YES
WARNING: You should add the boolean check kwarg to the run_command call.
         It currently defaults to false,
         but it will default to true in future releases of meson.
         See also: https://github.com/mesonbuild/meson/issues/9300
Library m found: YES
Library m found: YES
C++ compiler for the host machine: c++ (gcc 12.1.0 "c++ (Rev2, Built by MSYS2 project) 12.1.0")
C++ linker for the host machine: c++ ld.bfd 2.38
Library setupapi found: YES
Library hid found: YES
Library wsock32 found: YES
Library wsock32 found: YES
Found pkg-config: C:\msys64\mingw32\bin/pkg-config.EXE (1.8.0)
Run-time dependency glib-2.0 found: YES 2.72.2
Library wsock32 found: YES
Library dinput8 found: YES
Library dxguid found: YES
Run-time dependency zlib found: YES 1.2.12
Library dinput8 found: YES
Library dxguid found: YES
Library regex found: YES
Library wsock32 found: YES
Library setupapi found: YES
Dependency zlib found: YES 1.2.12 (cached)
Dependency zlib found: YES 1.2.12 (cached)

src/daemon/meson.build:120:0: ERROR: run_target got unknown keyword arguments "dependencies"
`

**Comment by eric-akira** [comment]
Are two steam controllers supported on the windows version?
I'm running pre-release 4.9.12. Tried 2 controllers + 2 dongles, 2 controllers + 1 dongle, 1 controller with cable + 1 controller with dongle.
Don´t know how to provide more info, but I will be glad to help =)

**Comment by kauron** [comment]
Hi there, I've just tried the latest prerelease (0.4.9.12) in Windows 10, and I get the following error when launching `sc-controller-console.exe`:

```
[I] GUI          Starting SC Controller GUI (forced console) v0.4.5...
[D] GUI          Initializing python...
[d] GUI          Python home: [FOLDER]\sc-controller-0.4.9.12-win32-pre-release
[d] GUI          Python path: [FOLDER]\sc-controller-0.4.9.12-win32-pre-release\python;[FOLDER]\sc-controller-0.4.9.12-win32-pre-release\python\scc\lib;[FOLDER]\sc-controller-0.4.9.12-win32-pre-release\lib\python2.7[FOLDER]\sc-controller-0.4.9.12-win32-pre-release\lib\python2.7\lib-dynload;[FOLDER]\sc-controller-0.4.9.12-win32-pre-release\lib\python2.7\site-packages
[E] GUI          Failed to initialize gui_loader.py:
Traceback (most recent call last):
  File "[FOLDER]\sc-controller-0.4.9.12-win32-pre-release\python/gui_loader.py", line 19, in <module>
    glades = os.path.join(get_share_path(), "glade")
  File "[FOLDER]\sc-controller-0.4.9.12-win32-pre-release\python/scc/paths.py", line 117, in get_share_path
    return lib_bindings.scc_get_share_path().decode("utf-8")
  File "[FOLDER]\sc-controller-0.4.9.12-win32-pre-release\lib\python2.7/encodings/utf_8.py", line 16, in decode
    return codecs.utf_8_decode(input, errors, True)
UnicodeDecodeError: 'utf8' codec can't decode byte 0xed in position 14: invalid continuation byte
```

I don't know if you prefer me to file a separate issue or want to have all of them together in the same thread.

**Comment by kozec** [comment]
> UnicodeDecodeError: 'utf8' codec can't decode byte 0xed in position 14: invalid continuation byte

Do you have any accented character in [FOLDER] by a chance? Error like this is kinda hard to fix, but easy to workaround, just move [FOLDER] to some place where there are no non-ascii characters in the path.

**Comment by kozec** [comment]
> Are two steam controllers supported on the windows version?

Only theoretically. Have you occured any issue, or it just plainly doesn't work?

**Comment by kauron** [comment]
> > UnicodeDecodeError: 'utf8' codec can't decode byte 0xed in position 14: invalid continuation byte
> 
> Do you have any accented character in [FOLDER] by a chance? Error like this is kinda hard to fix, but easy to workaround, just move [FOLDER] to some place where there are no non-ascii characters in the path.

The path is english letters (A-Z, a-z, 0-9, '-', '_', '.') only, I even tried placing it in the root of a drive (D:/) and still get the same error. Maybe it is trying to access a temp or appdata folder from the user, which does have an accent in the path.

**Comment by smvicente** [comment]
> > Are two steam controllers supported on the windows version?
> 
> Only theoretically. Have you occured any issue, or it just plainly doesn't work?

I can also confirm 2 controllers at the same time doesn't seem to be working. The second one still acts as a mouse and keyboard, but only the first one connected is able to change modes and control the OSD when pressing the Steam button.

Thank you for all the effort in any case!

**Comment by Ryochan7** [comment]
I have still been working on a different Steam Controller mapper for Windows although changes have been less frequent lately. A new build has just been pushed.

https://github.com/Ryochan7/SteamControllerTest/releases/tag/v0.0.13

**Comment by smvicente** [comment]
Hi @Ryochan7! do your Steam Controller mapper work with 2 controllers at the same time?

Thank you for that and for your Python 3 version of the sc-controller for Linux.

**Comment by Ryochan7** [comment]
Earlier versions of the mapper would have had issues with multiple controller support. Fixed the issue in my DS4 focused mapper and carried the list changes over to the Steam Controller mapper. Version 0.0.16 should work fine for multiple Steam Controllers.

**Comment by virgoparna** [comment]
Is https://github.com/kozec/sc-controller/wiki/Running-SC-Controller-on-Windows missing some crucial part? Following it and having downloaded sc-controller-0.4.9.10-win32-pre-release.zip scc-daemon.exe works and also osd (virtual controller appears and works just fine with X-Box controller profile). 
But sc-controller.exe fails to launch. And sc-controller-console.exe shows:

```
[E] GUI          Failed to initialize gui_loader.py:
Traceback (most recent call last):
  File "C:\bin\scc\python/gui_loader.py", line 4, in <module>
    gi.require_version('Gtk', '3.0')
  File "C:\bin\scc\lib\python2.7\site-packages/gi/__init__.py", line 129, in require_version
    raise ValueError('Namespace %s not available' % namespace)
ValueError: Namespace Gtk not available

```
I even changed path in terminal window before executing sc-controller-console.exe to minimal path 
C:\Windows\System32;C:\Windows;C:\Windows\System32\wbem 
to exclude local Python 3.11 install from path.
